### ISSUE TITLE
How do I build this network with mlpack?

### ISSUE THREAD

**Issue by RabbleRousy** [mlpack/mlpack#3103]
Hello everyone,
I am trying to train a Effectively Updatable Neural Network ([NNUE](https://www.chessprogramming.org/NNUE)) for Chess Position Evaluation. It is a fairly simple network, consisting only of 4 Layers, with the largest being the input size of 81920.
I have already built the network capable of calculating a forward pass in my engine according to [this tutorial](https://github.com/glinscott/nnue-pytorch/blob/master/docs/nnue.md#forward-pass-implementation), but I wanna use mlpack to actually train a net and use its weights.

But the net has one special feature which I struggle to implement using mlpack. The input features consist of two parts, each of size 40960, which are then both multiplied by the SAME matrix to form the first hidden layer, which is then 2 x 256 in size.

![Netzwerkstruktur](https://user-images.githubusercontent.com/61968673/145305948-2afd07f6-ece9-4f81-9f60-db1d0fc2066e.jpeg)
Each _Li_ is a fully connected `Linear` layer followed by clipped ReLU (I am currently using regular `ReLULayer` because I couldn't find a clipped ReLU layer in mlpack). I tried stitching together the architechture of the first part in mlpack using the `SubView`, `Sequential` and `Concat` Layers:

![Netzwerkstruktur mlpack](https://user-images.githubusercontent.com/61968673/145306302-af746899-5248-40f2-b069-da39fa8f0d38.jpeg)

Here is the code:

	mlpack::ann::FFN<mlpack::ann::MeanSquaredError<>> network;

	mlpack::ann::Sequential<> firstHalf(2 * N, M);
	firstHalf.Add<mlpack::ann::Subview<>>(1, 0, N - 1, 0, 0);
	firstHalf.Add<mlpack::ann::Linear<>>(2 * N, M);
	firstHalf.Add<mlpack::ann::ReLULayer<>>();

	mlpack::ann::Sequential<> secondHalf(2 * N, M);
	secondHalf.Add<mlpack::ann::Subview<>>(1, N, 2 * N - 1, 0, 0);
	secondHalf.Add<mlpack::ann::Linear<>>(2 * N, M);
	secondHalf.Add<mlpack::ann::ReLULayer<>>();

	mlpack::ann::Concat<> concat(2 * N, 2 * M);
	concat.Add<mlpack::ann::Sequential<>>(firstHalf);
	concat.Add<mlpack::ann::Sequential<>>(secondHalf);

	// 2 x L0
	network.Add<mlpack::ann::Concat<>>(concat);
	// L1
	network.Add<mlpack::ann::Linear<> >(2 * M, K);
	network.Add<mlpack::ann::ReLULayer<> >();
	// L2
	network.Add<mlpack::ann::Linear<> >(K, K);
	network.Add<mlpack::ann::ReLULayer<> >();
	// L3
	network.Add<mlpack::ann::Linear<> >(K, 1);
	network.Add<mlpack::ann::ReLULayer<> >();

It compiles (took me long enough), but when running and trying to train this net I get a Debug Assertion Error Message:
`Expression: front() called on empty vector`

**I am sorry** if this is all probably completely wrong and stupid, but I am really new to implementing anything ML related and just did not get any further with the sparse documentation of `SubView`, `Sequential` and `Concat`. I would be incredibly grateful for any tips on how to construct a network as pictured above in mlpack. Thanks!

**Comment by zoq** [comment]
That is an interesting usecase, the architecture looks like a correct solution to me, perhaps you can provide a really simple ready to use code snippet to reproduce the issue you see?

Another solution I could think of is to implement a custom layer based on the linear layer, it would be pretty similar to the existing solution except that we split the input before the forward pass and multiple each with the same matrix. I could write some pseudo code if you are interested in that option.

**Comment by RabbleRousy** [comment]
Hi @zoq , thanks for your answer! Wow, I was really not expecting to hear that my approach looks alright. If that‘s the case, I will happily give you a minimal complete example today, just give me some minutes ^^

I thought about the custom layer approach too. Before I found `Sequential`, `SubView` and `Concat` Layers I tried building the net by extending `FFN` class and overwriting the `forward()` method, but I struggled and didn‘t get it to work. It‘s probably the wrong approach and using one custom layer instead of a whole custom network class would be the way to go. 
But I wanted to try to get this method working before looking into custom layers, as I am kinda afraid of writing the Backward pass myself. So if you don‘t mind, I would love to see your take on that as pseudo-code!:)

Thanks again!!

**Comment by RabbleRousy** [comment]
Minimal, complete, reproducable example:

    #include <mlpack/core.hpp>
    #include <mlpack/methods/ann/ffn.hpp>
    #include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
    
    int main() {
	    const int N = 40960;
	    const int M = 256;
	    const int K = 32;
    
	    arma::sp_mat sparseMatrix;
	    // Only 50 training examples, as I haven't converted more to the correct format
	    sparseMatrix.load("trainData.csv", arma::coord_ascii);
	    sparseMatrix = sparseMatrix.t();
	    // Not sure if I could continue to use sp_mat, would be great
	    arma::mat data = (arma::mat)sparseMatrix.submat(0, 0, sparseMatrix.n_rows - 2, sparseMatrix.n_cols - 1);
    
	    // Cut the labels from the last row of the trainingData
	    arma::mat labels = (arma::mat)sparseMatrix.submat(sparseMatrix.n_rows - 1, 0, sparseMatrix.n_rows - 1, sparseMatrix.n_cols - 1);
    
	    mlpack::ann::FFN<mlpack::ann::MeanSquaredError<>> network;
    
	    mlpack::ann::Sequential<> firstHalf(2 * N, M);
	    firstHalf.Add<mlpack::ann::Subview<>>(1, 0, N - 1, 0, 0);
	    firstHalf.Add<mlpack::ann::Linear<>>(2 * N, M);
	    firstHalf.Add<mlpack::ann::ReLULayer<>>();
    
	    mlpack::ann::Sequential<> secondHalf(2 * N, M);
	    secondHalf.Add<mlpack::ann::Subview<>>(1, N, 2 * N - 1, 0, 0);
	    secondHalf.Add<mlpack::ann::Linear<>>(2 * N, M);
	    secondHalf.Add<mlpack::ann::ReLULayer<>>();
    
	    mlpack::ann::Concat<> concat(2 * N, 2 * M);
	    concat.Add<mlpack::ann::Sequential<>>(firstHalf);
	    concat.Add<mlpack::ann::Sequential<>>(secondHalf);
    
	    // Add the concatenation to the network
	    network.Add<mlpack::ann::Concat<>>(concat);
	    // L1
	    network.Add<mlpack::ann::Linear<> >(2 * M, K);
	    network.Add<mlpack::ann::ReLULayer<> >();
	    // L2
	    network.Add<mlpack::ann::Linear<> >(K, K);
	    network.Add<mlpack::ann::ReLULayer<> >();
	    // L3
	    network.Add<mlpack::ann::Linear<> >(K, 1);
    
	    network.Train(data, labels);
	    return 0;
    }
This is `main.cpp`, and I run the application in Debug mode. Error is the one described above.
`trainData.csv`, together with this example file, can be found on this [git branch](https://gitlab.fsi.hochschule-trier.de/hetzers/chess/-/tree/mlpack-testing).

**Comment by RabbleRousy** [comment]
Here is my solution following @zoq's recommendation of implementing a custom layer based on `Linear`:

    // Custom Forward pass splits the input in half and applies the weights and biases separately before joining the two 
    halves back together
    template<typename InputDataType, typename OutputDataType,
        typename RegularizerType>
        template<typename eT>
    void LinearSplit<InputDataType, OutputDataType, RegularizerType>::Forward(
        const arma::Mat<eT>& input, arma::Mat<eT>& output)
    {
        arma::mat i_firstHalf = input.submat(0, 0, inSize / 2 - 1, input.n_cols - 1);
        arma::mat i_secondHalf = input.submat(inSize / 2, 0, inSize - 1, input.n_cols - 1);
        arma::mat o_firstHalf = weight * i_firstHalf;
        o_firstHalf.each_col() += bias;
        arma::mat o_secondHalf = weight * i_secondHalf;
        o_secondHalf.each_col() += bias;
        output = arma::join_cols(o_firstHalf, o_secondHalf);
    }

    // Backward pass also calculates two halves of gradient vector separately
    // Never called in my usecase because LinearSplit is first layer
    template<typename InputDataType, typename OutputDataType,
        typename RegularizerType>
        template<typename eT>
    void LinearSplit<InputDataType, OutputDataType, RegularizerType>::Backward(
        const arma::Mat<eT>& /* input */, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
    {
        arma::mat error_firstHalf = gy.submat(0, 0, outSize / 2 - 1, gy.n_cols - 1);
        arma::mat error_secondHalf = gy.submat(outSize / 2, 0, outSize - 1, gy.n_cols - 1);
        g = arma::join_cols(weight.t() * error_firstHalf, weight.t() * error_secondHalf);
    }

    // Evaluate twice, using one Gradient from the first half, and another one from the second
    // Half of gradient matrix stays unused at is has double size of our actual weights
    template<typename InputDataType, typename OutputDataType,
        typename RegularizerType>
        template<typename eT>
    void LinearSplit<InputDataType, OutputDataType, RegularizerType>::Gradient(
        const arma::Mat<eT>& input,
        const arma::Mat<eT>& error,
        arma::Mat<eT>& gradient)
    {
        arma::mat i_firstHalf = input.submat(0, 0, inSize / 2 - 1, input.n_cols - 1);
        arma::mat i_secondHalf = input.submat(inSize / 2, 0, inSize - 1, input.n_cols - 1);
        arma::mat error_firstHalf = error.submat(0, 0, outSize / 2 - 1, error.n_cols - 1);
        arma::mat error_secondHalf = error.submat(outSize / 2, 0, outSize - 1, error.n_cols - 1);

        gradient.submat(0, 0, weight.n_elem - 1, 0) = arma::vectorise(
            error_firstHalf * i_firstHalf.t());
        gradient.submat(weight.n_elem, 0, gradient.n_elem - 1, 0) =
            arma::sum(error_firstHalf, 1);
        regularizer.Evaluate(weights, gradient);

        gradient.submat(0, 0, weight.n_elem - 1, 0) = arma::vectorise(
            error_secondHalf * i_secondHalf.t());
        gradient.submat(weight.n_elem, 0, gradient.n_elem - 1, 0) =
            arma::sum(error_secondHalf, 1);
        regularizer.Evaluate(weights, gradient);
    }
All other methods are exactly the same as in `Linear`.
This seems to work, atleast I can train the network and the loss drops to <0.002. Unfortunately, the network seems to be learning the median of all the training labels, but I will open another issue for that if I struggle to solve that.
